File: leetcode/92/92.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        ListNode *front = head, *tail, *preFront;
        while(m > 1) {
            preFront = front;
            front = front->next;
            m--;
            n--;
        }
        tail = front->next;
        ListNode *nextTail, *preTail = front;
        while(n > 1) {
            nextTail = tail->next;
            tail->next = preTail;
            preTail = tail;
            tail = nextTail;
            n--;
        }
        if(front == head) {
            head = preTail;
            front->next = tail;
        }
        else {
            preFront->next = preTail;
            front->next = tail;            
        }
        return head;
    }
};

File: leetcode/92/92.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if(left == right) {
            return head;
        }
        ListNode preHead = new ListNode();
        preHead.next = head;
        ListNode leftNode = preHead, rightNode, preLeftNode, pre, nextRightNode = null;
        int nextLoc = 1;
        while(nextLoc < left) {
            leftNode = leftNode.next;
            ++nextLoc;
        }
        preLeftNode = leftNode;
        leftNode = leftNode.next;
        pre = leftNode;
        rightNode = leftNode.next;
        int curLoc = nextLoc + 1;
        while(curLoc <= right) {
            nextRightNode = rightNode.next;
            rightNode.next = pre;
            pre = rightNode;
            rightNode = nextRightNode;
            ++curLoc;
        }
        preLeftNode.next = pre;
        leftNode.next = rightNode;
        return preHead.next;
    }
}